File: tests/importer/tflite_/basic/test_split.py
# Copyright 2019-2021 Canaan Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""System test: test concat"""
# pylint: disable=invalid-name, unused-argument, import-outside-toplevel

import pytest
import tensorflow as tf
import numpy as np
from tflite_test_runner import TfliteTestRunner


def _make_module(input_shape, num_split, axis):
    class SplitModule(tf.Module):
        def __init__(self):
            super(SplitModule).__init__()

        @tf.function(input_signature=[tf.TensorSpec(input_shape, tf.float32)])
        def __call__(self, x):
            return tf.split(x, num_or_size_splits=num_split, axis=axis)
    return SplitModule()


cases = [
    ([4], 2, 0),
    ([9], 3, -1),
    ([16, 8], 4, 0),
    ([16, 16], 4, 1),
    ([16, 32], 4, -1),
    ([10, 32], 5, -2),
    ([16, 8, 9], 8, 0),
    ([1, 8, 9], 2, 1),
    ([1, 8, 9], 3, 2),
    ([1, 8, 18], 3, -1),
    ([1, 3, 18], 3, -2),
    ([6, 8, 18], 3, -3),
    ([32, 8, 18, 7], 4, 0),
    ([32, 8, 18, 7], 2, 1),
    ([32, 8, 18, 7], 1, 2),
    ([32, 8, 18, 7], 7, 3),
    ([32, 8, 18, 2], 2, -1),
    ([32, 8, 18, 2], 3, -2),
    ([32, 8, 18, 2], 4, -3),
    ([36, 8, 18, 2], 6, -4),


]


@pytest.mark.parametrize('in_shapes,num_split,axis', cases)
def test_split(in_shapes, num_split, axis, request):
    module = _make_module(in_shapes, num_split, axis)

    runner = TfliteTestRunner(request.node.name)
    model_file = runner.from_tensorflow(module)
    runner.run(model_file)


if __name__ == "__main__":
    pytest.main(['-vv', 'test_split.py'])

File: tests/kernels/test_split.cpp
/* Copyright 2019-2021 Canaan Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "kernel_test.h"
#include <gtest/gtest.h>
#include <iostream>
#include <nncase/kernels/stackvm/tensor_ops.h>
#include <nncase/runtime/datatypes.h>
#include <nncase/runtime/runtime_tensor.h>
#include <nncase/runtime/simple_types.h>
#include <nncase/runtime/stackvm/opcode.h>
#include <ortki/operators.h>

using namespace nncase;
using namespace nncase::runtime;
using namespace ortki;

#define TEST_CASE_NAME "test_split"

class SplitTest : public KernelTest,
                  public ::testing::TestWithParam<std::tuple<int>> {
  public:
    void SetUp() override {
        READY_SUBCASE()

        auto l_shape = GetShapeArray("lhs_shape");
        auto value = GetNumber("axis");
        auto typecode = GetDataType("lhs_type");

        input =
            hrt::create(typecode, l_shape, host_runtime_tensor::pool_cpu_only)
                .expect("create tensor failed");
        init_tensor(input);

        axis_value = value;
        int64_t axis_array[] = {axis_value};
        axis = hrt::create(dt_int64, {1},
                           {reinterpret_cast<std::byte *>(axis_array),
                            sizeof(axis_array)},
                           true, host_runtime_tensor::pool_cpu_only)
                   .expect("create tensor failed");
    }

    void TearDown() override { CLEAR_SUBCASE() }

  protected:
    runtime_tensor input;
    int64_t axis_value;
    runtime_tensor axis;
};

INSTANTIATE_TEST_SUITE_P(Split, SplitTest,
                         testing::Combine(testing::Range(0, MAX_CASE_NUM)));

TEST_P(SplitTest, Split) {
    auto l_ort = runtime_tensor_2_ort_tensor(input);

    // expected
    size_t size = 0;
    int64_t sections_array[] = {2, 2};
    auto sections = hrt::create(dt_int64, {2},
                                {reinterpret_cast<std::byte *>(sections_array),
                                 sizeof(sections_array)},
                                true, host_runtime_tensor::pool_cpu_only)
                        .expect("create tensor failed");

    auto output_ort1 = tensor_seq_get_value(
        ortki_Split(l_ort, runtime_tensor_2_ort_tensor(sections), axis_value),
        0);
    void *ptr_ort1 = tensor_buffer(output_ort1, &size);
    dims_t shape1(tensor_rank(output_ort1));
    tensor_shape(output_ort1, reinterpret_cast<int64_t *>(shape1.data()));
    auto expected1 =
        hrt::create(input.datatype(), shape1,
                    {reinterpret_cast<std::byte *>(ptr_ort1), size}, true,
                    host_runtime_tensor::pool_cpu_only)
            .expect("create tensor failed");

    auto output_ort2 = tensor_seq_get_value(
        ortki_Split(l_ort, runtime_tensor_2_ort_tensor(sections), axis_value),
        1);
    void *ptr_ort2 = tensor_buffer(output_ort2, &size);
    dims_t shape2(tensor_rank(output_ort2));
    tensor_shape(output_ort2, reinterpret_cast<int64_t *>(shape2.data()));
    auto expected2 =
        hrt::create(input.datatype(), shape2,
                    {reinterpret_cast<std::byte *>(ptr_ort2), size}, true,
                    host_runtime_tensor::pool_cpu_only)
            .expect("create tensor failed");

    runtime_tensor expected[] = {expected1, expected2};
    typecode_t type[] = {dt_float32, dt_float32};

    // actual
    auto output =
        kernels::stackvm::split(input.impl(), axis.impl(), sections.impl())
            .expect("split failed");
    tuple actual(output.as<tuple>().expect("as tensor failed"));

    [[maybe_unused]] auto result = check_tuple_output(expected, type, output);
}

int main(int argc, char *argv[]) {
    READY_TEST_CASE_GENERATE()
    FOR_LOOP(lhs_shape, i)
    FOR_LOOP(axis, j)
    FOR_LOOP(lhs_type, k)
    SPLIT_ELEMENT(lhs_shape, i)
    SPLIT_ELEMENT(axis, j)
    SPLIT_ELEMENT(lhs_type, k)
    WRITE_SUB_CASE()
    FOR_LOOP_END()
    FOR_LOOP_END()
    FOR_LOOP_END()

    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}